### PR TITLE
first PR heap jemalloc analysis

### PR DESCRIPTION
**Pull request by soez**

### CODE REVIEW

**Comment by radare** [comment]
```

  ^
In file included from /home/travis/build/radare/radare2/libr/include/r_heap_jemalloc.h:4:0,
                 from cmd_debug.c:14,
                 from cmd.c:76:
/home/travis/build/radare/radare2/libr/include/r_jemalloc/internal/jemalloc_internal.h:153:23: fatal error: stdatomic.h: No such file or directory
 #include <stdatomic.h>
                       ^
compilation terminated.
make[3]: *** [cmd.o] Error 1
make[3]: *** Waiting for unfinished jobs....
```


doesnt builds

File: libr/core/linux_heap_jemalloc.c [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+				free (path);
+				return true;
+			} 
+		}
+	}
+}
+
+static void GH(jemalloc_get_runs)(RCore *core, const char *input) {
+    switch (input[0]) {
+        case ' ':
```
Comment (radare) [inline libr/core/linux_heap_jemalloc.c]: cases indented at the same level as switch

File: libr/core/linux_heap_jemalloc.c [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+                eprintf ("run: %"PFMT64x"\n", miscelm->run);
+                eprintf ("miscelm: %"PFMT64x"\n", &miscelm);
+                eprintf ("miscelm->run: %"PFMT64x"\n", &miscelm->run);
+                sleep(1);
+           } 
+            eprintf ("offset: %"PFMT64x"\n", offset);
+            eprintf ("run @ %"PFMT64x"\n", misc + offset);
+         */     
+        }
+            break;
```
Comment (radare) [inline libr/core/linux_heap_jemalloc.c]: wrong indent

File: libr/core/linux_heap_jemalloc.c [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+            arena_chunk_t *c = R_NEW0 (arena_chunk_t);
+            arena_chunk_map_misc_t *miscelm = R_NEW0 (arena_chunk_map_misc_t);
+            arena_chunk_map_bits_t *bits = R_NEW0 (arena_chunk_map_bits_t);
+            int flag = 0;
+
+            input +=1;
+            chunk = strstr (input, "0x") ? (GHT)strtol (input, NULL, 0) : (GHT)strtol (input, NULL, 16);
+        
+            if (GH(r_resolve_jemalloc)(core, "je_map_misc_offset", &map_misc_offset)) {
+                r_core_read_at (core, map_misc_offset, (ut8*)&map_misc_offset, sizeof (GHT));
```
Comment (radare) [inline libr/core/linux_heap_jemalloc.c]: use tabs, not spaces

**Comment by radare** [comment]
are all those include files needed? can we reduce the amount of them?

**Comment by XVilka** [comment]
While you're on it - what about moving heap analysis from the libr/core to libr/debug, where it makes more sense?

**Comment by radare** [comment]
Yes makes more sense to be in debug. But this should change in a different pr. And im not sure if debug is the place because this may work with non debugging environments too. It fits better into anal imho

File: libr/core/cconfig.c [hunk 4]
```diff
@@ -1915,6 +1915,19 @@ static int cb_linesabs(void *user, void *data) {
...
 
+static int cb_malloc(void *user, void *data) {
+ 	RCore *core = (RCore*) user;
+ 	RConfigNode *node = (RConfigNode*) data;
+
+ 	if (node->value){
+ 		if (!strcmp ("jemalloc", node->value) || !strcmp ("glibc", node->value)) {
+ 			core->dbg->malloc = data;
+ 		}
+
```
Comment (radare) [inline libr/core/cconfig.c]: empty line

File: libr/include/r_jemalloc/internal/extent.h [hunk 5]
```diff
@@ -0,0 +1,275 @@
+/******************************************************************************/
```
Comment (radare) [inline libr/include/r_jemalloc/internal/extent.h]: i dont think we want to expose jemalloc include files to the public, put those inside libr/core/p or somehwere else only available for the implementation. but i think we shuold move all this stuff into shlr/heapanal or libr/anal/heap.c

**Comment by radare** [comment]
tweaked and merged! still lot to do be done, but enough to go forward!

thanks!

**Pull request closed without merging**